File: backend/Semsar-Backend/Domain/Entities/Property.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using Domain.Enums;

namespace Domain.Entities
{
    public class Property : ISoftDelete, IHasPublicKey
    {
        public int Id { get; set; }
        public Guid PublicId { get; set; } = Guid.CreateVersion7();
        public string PublicKey { get; set; } = string.Empty;

        // ===== Multilingual =====
        [Required]
        [MaxLength(200)]
        public string TitleEn { get; set; } = null!;

        [Required]
        [MaxLength(200)]
        public string TitleAr { get; set; } = null!;

        [MaxLength(8000)]
        public string DescriptionEn { get; set; } = null!;

        [MaxLength(8000)]
        public string DescriptionAr { get; set; } = null!;

        // ===== Pricing =====
        [Range(0, double.MaxValue)]
        public decimal Price { get; set; }

        public decimal? RentPerMonth { get; set; }

        [MaxLength(10)]
        public string Currency { get; set; } = "EGP";

        // ===== Classification =====
        public PropertyType PropertyType { get; set; }
        public PropertyListingType ListingType { get; set; }

        [Required]
        [MaxLength(200)]
        public string Location { get; set; } = null!;

        public string? LocationAr { get; set; }

        public int? LocationId { get; set; }
        public Location? LocationEntity { get; set; }

        [MaxLength(50)]
        public string Code { get; set; } = null!;

        // routing
        [MaxLength(450)]
        public string Slug { get; set; } = null!;

        public bool SlugIsAuto { get; set; } = true;

        [MaxLength(5)]
        public string SlugLanguage { get; set; } = "en";

        [MaxLength(200)]
        public string? SeoTitle { get; set; }

        [MaxLength(300)]
        public string? SeoDescription { get; set; }

        [MaxLength(200)]
        public string? SeoTitleAr { get; set; }

        [MaxLength(300)]
        public string? SeoDescriptionAr { get; set; }

        [MaxLength(500)]
        public string? SeoKeywords { get; set; }

        [MaxLength(500)]
        public string? SeoKeywordsAr { get; set; }

        [MaxLength(1000)]
        public string CanonicalUrl { get; set; } = string.Empty;

        public DateTime MetaGeneratedAt { get; set; } = DateTime.UtcNow;
        public int MetaVersion { get; set; } = 1;

        public bool IsFeatured { get; set; } = false;
        public bool IsRecommended { get; set; } = false;
        public int SortOrder { get; set; } = 0;

        [Range(0, double.MaxValue)]
        public double Size { get; set; }

        public int Bedrooms { get; set; } = 0;
        public int Bathrooms { get; set; } = 0;
        public int? Floor { get; set; }
        public int? TotalFloors { get; set; }
        public bool IsFurnished { get; set; } = false;
        public PropertyView View { get; set; } = PropertyView.Unknown;

        public List<string> Features { get; set; } = new();
        public List<string> FeaturesAr { get; set; } = new();
        public ICollection<PropertyFeature> PropertyFeatures { get; set; } = new List<PropertyFeature>();

        public int? ContactId { get; set; }
        public ContactInfo? Contact { get; set; }

        public ICollection<PropertyImage>? Images { get; set; }
        public ICollection<PropertyVideo>? Videos { get; set; }
        public ICollection<PropertyInstallmentPlan>? Installments { get; set; } = new List<PropertyInstallmentPlan>();
        public RentalDetails? RentalDetails { get; set; }

        [MaxLength(200)]
        public string? DeliveryText { get; set; }
        [MaxLength(200)]
        public string? DeliveryTextAr { get; set; }
        public ConstructionStatus? ConstructionStatus { get; set; }
        [MaxLength(50)]
        public string? AvailabilityStatus { get; set; } = "Available";
        public OwnershipType? OwnershipType { get; set; }
        public int InquiryCount { get; set; } = 0;
        public int FavoriteCount { get; set; } = 0;
        [MaxLength(1000)]
        public string? VirtualTourUrl { get; set; }
        public List<string>? HighlightsAr { get; set; }
        public List<string>? NearbyPlaces { get; set; }
        public List<string>? NearbyPlacesAr { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? UpdatedAt { get; set; }
        public bool IsDeleted { get; set; } = false;
        public int ViewCount { get; set; } = 0;
        public int ContactClickCount { get; set; } = 0;
        public byte[] RowVersion { get; set; } = Array.Empty<byte>();

        public override bool Equals(object? obj)
        {
            if (obj is not Property other) return false;
            if (ReferenceEquals(this, other)) return true;
            return Id != 0 && Id == other.Id;
        }

        public override int GetHashCode()
        {
            return Id.GetHashCode();
        }
    }
}

File: frontend/semsar-web/src/types/property.ts
export type PropertyType = 'Studio' | '1BR' | '2BR' | 'Villa';

export type PropertyCategory =
  | 'Apartment' | 'Villa' | 'Office' | 'Shop' | 'Land' | 'Building' | 'Penthouse' | 'Duplex' | 'Studio' | 'Chalet' | 'Other';

export type PropertyListingType = 'Resale' | 'Rental' | 'Project';

export type PropertyStatus = 'Available' | 'Reserved' | 'Sold';

export const PROPERTY_VIEWS = ['Sea', 'Pool', 'Garden', 'Street', 'City', 'Golf', 'Park', 'Lake', 'Mountain', 'BackView', 'SideSeaView', 'PoolSeaView', 'SeaPoolView', 'Unknown'] as const;
export type PropertyView = (typeof PROPERTY_VIEWS)[number];

export const FINISHING_TYPES = ['FullyFinished', 'SemiFinished', 'CoreAndShell'] as const;
export type FinishingType = (typeof FINISHING_TYPES)[number];

export type ListingType = 'sale' | 'rent';

export type AvailabilityStatus = 'Available' | 'Reserved' | 'SoldOut';

export type ConstructionStatus = 'Planned' | 'UnderConstruction' | 'NearDelivery' | 'Delivered';

export type OwnershipType = 'Freehold' | 'GreenContract' | 'Usufruct' | 'SharedOwnership' | 'Other';

export const OWNERSHIP_TYPES: OwnershipType[] = ['Freehold', 'GreenContract', 'Usufruct', 'SharedOwnership', 'Other'];

export const CONSTRUCTION_STATUSES: ConstructionStatus[] = ['Planned', 'UnderConstruction', 'NearDelivery', 'Delivered'];

export interface NearbyPlace {
  name: string;
  nameAr?: string;
  distance: number;
  icon?: string;
}

export interface Variant {
  id: number;
  publicKey?: string;
  name: string;
  nameAr?: string;
  size: number;
  price: number;
  currency: string;
  rentPerMonth?: number;
  bedrooms: number;
  bathrooms: number;
  floor?: number;
  isFurnished: boolean;
  view?: string;
  unitNumber?: string;
  buildingNumber?: string;
  deliveryDate?: string;
  deliveryText?: string;
  finishingType?: string;
  hasBalcony: boolean;
  hasParking: boolean;
  images?: string[];
  floorPlanUrl?: string;
  availabilityStatus?: AvailabilityStatus;
  sortOrder: number;
  isActive: boolean;
  isFeatured?: boolean;
  isRecommended?: boolean;
  viewCount?: number;
  inquiryCount?: number;
  favoriteCount?: number;
}

export interface FinancingCalculation {
  variantPrice: number;
  downPaymentPercent: number;
  years: number;
  downPaymentAmount: number;
  remainingAmount: number;
  monthlyInstallment: number;
  currency: string;
}

export interface InstallmentPlan {
  paymentType?: 'Installment' | 'Cash';
  downPaymentPercent: number;
  discountPercent?: number;
  years: number;
  installmentMonths?: number;
  quarterlyAmount?: number;
  monthlyAmount?: number;
  isEnabled?: boolean;
  isDeleted?: boolean;
}

export interface VideoItem {
  id: number;
  url: string;
  publicId: string;
  thumbnailUrl?: string;
  sortOrder?: number;
}

export interface Property {
  id: string;
  rawId?: number;
  rawUnitId?: number;
  publicKey?: string;
  slug: string;
  propertyCode: string;

  titleEn: string;
  titleAr: string;
  descriptionEn: string;
  descriptionAr: string;
  title: string;
  description: string;

  type: PropertyType;
  propertyType: PropertyCategory;
  listingType: ListingType;
  listingTypeBackend: PropertyListingType;

  price: number;
  minPrice?: number;
  maxPrice?: number;
  rentPerMonth?: number;
  currency: string;

  location: string;
  locationAr?: string;
  size: number;
  minArea?: number;
  maxArea?: number;
  status: PropertyStatus;
  features: string[];
  featuresAr?: string[];

  bedrooms?: number;
  bathrooms?: number;
  floor?: number;
  totalFloors?: number;
  isFurnished?: boolean;
  view?: PropertyView;
  unitNumber?: string;
  buildingNumber?: string;
  deliveryDate?: string;
  deliveryText?: string;
  deliveryTextAr?: string;
  finishingType?: FinishingType;
  hasBalcony?: boolean;
  hasParking?: boolean;

  image: string;
  images: string[];
  videos?: VideoItem[];
  floorPlans?: string[];
  virtualTourUrl?: string;

  projectId: string | null;

  variants?: Variant[];
  unitType?: string;

  installments: InstallmentPlan[];
  installment?: InstallmentPlan;

  isFeatured?: boolean;
  highlights: string[];
  highlightsAr?: string[];
  nearbyPlaces?: NearbyPlace[];

  seoTitleEn?: string;
  seoTitleAr?: string;
  seoDescriptionEn?: string;
  seoDescriptionAr?: string;

  viewCount?: number;
  inquiryCount?: number;
  favoriteCount?: number;

  ownershipType?: OwnershipType;
  constructionStatus?: ConstructionStatus;
}

export interface Project {
  id: string;
  publicKey?: string;
  slug: string;

  nameEn: string;
  nameAr: string;
  descriptionEn: string;
  descriptionAr: string;

  name: string;
  description: string;

  location: string;
  locationAr?: string;
  developer?: string;
  image: string;
  images: string[];
  videos?: VideoItem[];
  highlights: string[];
  highlightsAr?: string[];
  startingPrice?: number;
  highestPrice?: number;
  propertyTypes?: string[];
  totalArea?: number;
  latitude?: number;
  longitude?: number;
  ownershipType?: string;
  nearbyPlaces?: string[];
  nearbyPlacesAr?: string[];
  unitCount?: number;
  totalAvailableUnits?: number;
  totalReservedUnits?: number;
  totalSoldUnits?: number;
  unitTypesCount?: number;
  units: Property[];
  deliveryText?: string;
  deliveryTextAr?: string;
  constructionStatus?: ConstructionStatus;
}

export interface AdvancedFilterState {
  minPrice: number | '';
  maxPrice: number | '';
  minSize: number | '';
  maxSize: number | '';
  locations: string[];
  types: PropertyCategory[];
}

export interface SiteSettings {
  whatsappNumber: string;
  phoneNumber: string;
  companyName: string;
  email?: string;
  socialLinks: {
    facebook?: string;
    instagram?: string;
    tiktok?: string;
  };
}

export interface BookingPayload {
  propertyId: number | null;
  unitId: number | null;
  name: string;
  phone: string;
  message?: string;
  source?: string;
  medium?: string;
  campaign?: string;
  term?: string;
  content?: string;
  landingPage?: string;
  firstVisitAt?: string;
  currentPage?: string;
  referrer?: string;
  userAgent?: string;
  pageViews?: number;
  sessionDuration?: number;
  lastReferrer?: string;
  visitHistory?: string;
}

export interface LandRequestPayload {
  name: string;
  phone: string;
  location: string;
  minPrice?: number;
  maxPrice?: number;
  minArea?: number;
  maxArea?: number;
  notes?: string;
  source?: string;
  medium?: string;
  campaign?: string;
  term?: string;
  content?: string;
  landingPage?: string;
  firstVisitAt?: string;
  currentPage?: string;
  referrer?: string;
  userAgent?: string;
  pageViews?: number;
  sessionDuration?: number;
  lastReferrer?: string;
  visitHistory?: string;
}
